python/cudf/cudf/_libxx/join.pyx

```diff
@@ -0,0 +1,150 @@
+from cudf._libxx.lib cimport *
+from cudf._libxx.table cimport Table, columns_from_ptr
+from cudf._libxx.includes.join cimport *
+from libcpp.vector cimport vector
+from libcpp.pair cimport pair
+
+from collections import OrderedDict
+
+cpdef join(Table lhs,
+           Table rhs,
+           object left_on,
+           object right_on,
+           object how,
+           object method,
+           bool left_index=False,
+           bool right_index=False
+           ):
+    """
+    Call libcudf++ join for full outer, inner and left joins.
+    Returns a list of tuples [(column, valid, name), ...]
+    """
+
+    cdef Table c_lhs = lhs
+    cdef Table c_rhs = rhs
+
+    num_inds_left = len(left_on) + left_index
+    num_inds_right = len(right_on) + right_index
+    cdef vector[int] left_on_ind
+    cdef vector[int] right_on_ind
+    left_on_ind.reserve(num_inds_left)
+    right_on_ind.reserve(num_inds_right)
+
+    cdef vector[pair[int, int]] columns_in_common
+
+    # Views might or might not include index
+    cdef table_view lhs_view
+    cdef table_view rhs_view
+
+    # the result columns are all the left columns (including common ones)
+    # + all the right columns (excluding the common ones)
+    result_col_names = []
+
+    for name in lhs._data.keys():
+        result_col_names.append(name)
+    for name in rhs._data.keys():
+        if name not in lhs._data.keys():
+            result_col_names.append(name)
+
+    # keep track of where the desired index column will end up
+    result_index_pos = None
+
+    if left_index or right_index:
+        # If either true, we need to process both indices as columns
+        lhs_view = c_lhs.view()
+        rhs_view = c_rhs.view()
+
+        left_join_cols = [lhs.index.name] + list(lhs._data.keys())
+        right_join_cols = [rhs.index.name] + list(rhs._data.keys())
+        if left_index and right_index:
+            left_on_idx = right_on_idx = 0
+            result_index_pos = 0
+            result_index_name = rhs.index.name
+
+        elif left_index:
+            left_on_idx = 0
+            right_on_idx = right_join_cols.index(right_on[0])
+            result_index_pos = len(left_join_cols)
+            result_index_name = rhs.index.name
+
+            # common col gathered from the left
+            common = result_col_names.pop(result_col_names.index(right_on[0]))
+            result_col_names = [common] + result_col_names
+        elif right_index:
+            right_on_idx = 0
+            left_on_idx = left_join_cols.index(left_on[0])
+            result_index_pos = 0
+            result_index_name = lhs.index.name
+
+        left_on_ind.push_back(left_on_idx)
+        right_on_ind.push_back(right_on_idx)
+
+        columns_in_common.push_back(pair[int, int](
```

Comment:
Do we know the size of this up front? If so can we reserve it?

Reply:
Should be possible, updates to follow.

Reply:
Done - we're using a set to store the tuples as we go now and converting to a c++ vector later.